### ISSUE TITLE
docs on readthedocs aren't showing some standard features

### ISSUE THREAD

**Issue by rickeylev** [bazelbuild/rules_python#1516]
Our docs on readthedocs seem to be missing a couple of things that usually show up:

* PR builds don't show a warning banner (https://docs.readthedocs.io/en/stable/pull-requests.html)
* The mentions of github aren't shown on pages and in the flyout menu. Compare with https://rules-testing.readthedocs.io/en/latest/index.html ("Edit on github" at the top right", the flyout menu has a "Github: view/edit" link)

Not sure why. I'm guessing it's because Bazel is handling everything instead of readthedocs. Usually with readthedocs, the config file specifies:
  * python version to use
  * requirements file to install
  * Sphinx config file being used
  * Specific formats to build

And this gets translated into various `pip install` and `sphinx-build` commands.

e.g. here's some of the build log from rules_testing, which uses the readthedocs.yml config file:

```
python -mvirtualenv $READTHEDOCS_VIRTUALENV_PATH
python -m pip install --upgrade --no-cache-dir pip setuptools
python -m pip install --upgrade --no-cache-dir sphinx readthedocs-sphinx-ext
python -m pip install --exists-action=w --no-cache-dir -r docs/requirements.txt
npm install -g @bazel/bazelisk
bazel run //docs:run_sphinx_build
cat docs/source/conf.py
python -m sphinx -T -E -b html -d _build/doctrees -D language=en . $READTHEDOCS_OUTPUT/html
python -m sphinx -T -E -b readthedocssinglehtmllocalmedia -d _build/doctrees -D language=en . $READTHEDOCS_OUTPUT/htmlzip
python -m sphinx -T -E -b latex -d _build/doctrees -D language=en . $READTHEDOCS_OUTPUT/pdf
cat latexmkrc
latexmk -r latexmkrc -pdf -f -dvi- -ps- -jobname=rules-testing -interaction=nonstopmode
```

Maybe it also has to do with the environment variables? I noticed that environment variables aren't passed through to the `actions.run` call for running Sphinx. Which is WAI as far as bazel is concerned, but maybe there is logic within one of the sphinx extensions that checks for those to activate special behavior.

**Comment by rickeylev** [comment]
Aha, I think I found the issue.

With the regular build process, rtd appends a bunch of content on the conf.py to do special stuff. The most interesting parts are:
 * It adds a bunch of values to `html_context`. This includes a READTHEDOCS variable, list of other versions, and other metadata.
 * It adds some extensions: `readthedocs_ext.readthedocs`, `readthedocs_ext.external_version_warning`

This comes from this template: https://github.com/rtfd/readthedocs.org/blob/main/readthedocs/doc_builder/templates/doc_builder/conf.py.tmpl

So, this might be interesting to replicate. Few ideas:

* Set `sphinx.configuration` in readthedocs.yml. Maybe if it knows where the conf file is, it'll run its modification.
* Go back to the non-customized build process. We can probably do everything in a post-build step and do some sort of "stub" build to satisfy the normal build process.
* There's a RTD api, so, theorectically, I suppose it might be possible to have the build process query the api to get some of the info. That seems like a lot of work, though.
* I also found this, which says some of what has to be done, but they also say they arne't going to support this trick: https://github.com/readthedocs/readthedocs.org/issues/9342

**Comment by aignas** [comment]
One thing that I noticed whilst playing with this setup is that the `Edit on Github` will lead to nowhere for the API reference, but that for the time being might be an acceptable behavior.

**Comment by rickeylev** [comment]
For posterity, here is a dump of the extra content that gets injected to the rules_testing config file. From some quick local testing, the `html_context["READTHEDOCS"] = True` setting is the primary control flag.  Some of it is compatibility code we can just entirely omit.

It should be pretty easy to add this to our config file and set the value based on if its a local vs RTD build via flags.

```
#          auto-created readthedocs.org specific configuration            #
###########################################################################


#
# The following code was added during an automated build on readthedocs.org
# It is auto created and injected for every build. The result is based on the
# conf.py.tmpl file found in the readthedocs.org codebase:
# https://github.com/rtfd/readthedocs.org/blob/main/readthedocs/doc_builder/templates/doc_builder/conf.py.tmpl
#
# Note: this file shouldn't rely on extra dependencies.

import importlib
import sys
import os.path

# Borrowed from six.
PY3 = sys.version_info[0] == 3
string_types = str if PY3 else basestring

from sphinx import version_info

# Get suffix for proper linking to GitHub
# This is deprecated in Sphinx 1.3+,
# as each page can have its own suffix
if globals().get('source_suffix', False):
    if isinstance(source_suffix, string_types):
        SUFFIX = source_suffix
    elif isinstance(source_suffix, (list, tuple)):
        # Sphinx >= 1.3 supports list/tuple to define multiple suffixes
        SUFFIX = source_suffix[0]
    elif isinstance(source_suffix, dict):
        # Sphinx >= 1.8 supports a mapping dictionary for multiple suffixes
        SUFFIX = list(source_suffix.keys())[0]  # make a ``list()`` for py2/py3 compatibility
    else:
        # default to .rst
        SUFFIX = '.rst'
else:
    SUFFIX = '.rst'

# Add RTD Static Path. Add to the end because it overwrites previous files.
if not 'html_static_path' in globals():
    html_static_path = []
if os.path.exists('_static'):
    html_static_path.append('_static')

# Define this variable in case it's not defined by the user.
# It defaults to `alabaster` which is the default from Sphinx.
# https://www.sphinx-doc.org/en/master/usage/configuration.html#confval-html_theme
html_theme = globals().get('html_theme', 'alabaster')

#Add project information to the template context.
context = {
    'html_theme': html_theme,
    'current_version': "latest",
    'version_slug': "latest",
    'MEDIA_URL': "https://media.readthedocs.org/",
    'STATIC_URL': "https://assets.readthedocs.org/static/",
    'PRODUCTION_DOMAIN': "readthedocs.org",
    'proxied_static_path': "/_/static/",
    'versions': [
    ("latest", "/en/latest/"),
    ("stable", "/en/stable/"),
    ("v0.5.0", "/en/v0.5.0/"),
    ("v0.4.0", "/en/v0.4.0/"),
    ("v0.3.0", "/en/v0.3.0/"),
    ("v0.2.0", "/en/v0.2.0/"),
    ("v0.1.0", "/en/v0.1.0/"),
    ("v0.0.6", "/en/v0.0.6/"),
    ("v0.0.4", "/en/v0.0.4/"),
    ],
    'downloads': [ 
    ("pdf", "//rules-testing.readthedocs.io/_/downloads/en/latest/pdf/"),
    ("html", "//rules-testing.readthedocs.io/_/downloads/en/latest/htmlzip/"),
    ],
    'subprojects': [ 
    ],
    'slug': 'rules-testing',
    'name': u'rules_testing',
    'rtd_language': u'en',
    'programming_language': u'other',
    'canonical_url': '',
    'analytics_code': 'None',
    'single_version': False,
    'conf_py_path': '/docs/source/',
    'api_host': 'https://readthedocs.org',
    'github_user': 'bazelbuild',
    'proxied_api_host': '/_',
    'github_repo': 'rules_testing',
    'github_version': 'master',
    'display_github': True,
    'bitbucket_user': 'None',
    'bitbucket_repo': 'None',
    'bitbucket_version': 'master',
    'display_bitbucket': False,
    'gitlab_user': 'None',
    'gitlab_repo': 'None',
    'gitlab_version': 'master',
    'display_gitlab': False,
    'READTHEDOCS': True,
    'using_theme': (html_theme == "default"),
    'new_theme': (html_theme == "sphinx_rtd_theme"),
    'source_suffix': SUFFIX,
    'ad_free': False,
    'docsearch_disabled': False,
    'user_analytics_code': '',
    'global_analytics_code': 'UA-17997319-1',
    'commit': 'b88ee3c9',
}

# For sphinx >=1.8 we can use html_baseurl to set the canonical URL.
# https://www.sphinx-doc.org/en/master/usage/configuration.html#confval-html_baseurl
if version_info >= (1, 8):
    if not globals().get('html_baseurl'):
        html_baseurl = context['canonical_url']
    context['canonical_url'] = None

if 'html_context' in globals():
    for key in context:
        if key not in html_context:
            html_context[key] = context[key]
else:
    html_context = context

# Add custom RTD extension
if 'extensions' in globals():
    # Insert at the beginning because it can interfere
    # with other extensions.
    # See https://github.com/rtfd/readthedocs.org/pull/4054
    extensions.insert(0, "readthedocs_ext.readthedocs")
else:
    extensions = ["readthedocs_ext.readthedocs"]

# Add External version warning banner to the external version documentation
if 'branch' == 'external':
    extensions.insert(1, "readthedocs_ext.external_version_warning")
    readthedocs_vcs_url = 'None'
    readthedocs_build_url = 'https://readthedocs.org/projects/rules-testing/builds/22213338/'

project_language = 'en'

# User's Sphinx configurations
language_user = globals().get('language', None)
latex_engine_user = globals().get('latex_engine', None)
latex_elements_user = globals().get('latex_elements', None)

# Remove this once xindy gets installed in Docker image and XINDYOPS
# env variable is supported
# https://github.com/rtfd/readthedocs-docker-images/pull/98
latex_use_xindy = False

chinese = any([
    language_user in ('zh_CN', 'zh_TW'),
    project_language in ('zh_CN', 'zh_TW'),
])

japanese = any([
    language_user == 'ja',
    project_language == 'ja',
])

if chinese:
    latex_engine = latex_engine_user or 'xelatex'

    latex_elements_rtd = {
        'preamble': '\\usepackage[UTF8]{ctex}\n',
    }
    latex_elements = latex_elements_user or latex_elements_rtd
elif japanese:
    latex_engine = latex_engine_user or 'platex'

# Make sure our build directory is always excluded
exclude_patterns = globals().get('exclude_patterns', [])
exclude_patterns.extend(['_build'])
```

**Comment by rickeylev** [comment]
After a bit of experimenting, it looks like there's two basic approaches:

1. Use build.jobs. To make this work, we have to create a "stub" project for RTD to build, then in a post_build step, copy the modified config file over and do our full build. Thankfully, the stub project just needs an empty conf.py and index.rst. This does mean it goes through a full "install python, sphinx, some other stuff, sphinx build, copy output" build, and then we delete all that and do it again. So the build itself looks a bit funny.

2. In our conf.py file, set a few additional `html_context` settings. It looks like we only need to set a few; most of the others look unused in the sphinx_rtd_theme theme.
  * conf_py_path, github_user, github_repo, github_version. These control the URL generated for the "view source" or "edit on github" links.
  * commit. This controls the commit shown at the bottom of the page

The `html_context["READTHEDOCS"]` setting isn't strictly necessary. It affects:
 * Showing the RTD flyout _locally_, but not on RTD builds. RTD builds inject the flyout onto the page. 
 * The theme has some conditional for preventing double inclusion of analytic ids. But we don't use those, so I don't think we have to worry about it. 

The rest of the settings appear to be unused, deprecated, or not of interest.

I haven't looked into the two extensions it adds, yet. The external warning banner sounds self-explanatory. I'm not sure what readthedocs_ext.readthedocs is doing. If they have to be _conditionally_ added to the extensions (which seems likely, given the logic for adding the external-warning extension), then it's easy to add a `--sphinxdocs//sphinx:extra_env` flag to control the `env` arg of the sphinx invocation. (An alternative would be to pass conf.py through an `expand_template()` phase, but that would still require _some_ flag to have the value to pass into expand_template).
